crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in its per-module free list.
+    fn update_per_module_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Affinity {
+                ref mut per_module_index,
+                ..
+            } => {
+                *per_module_index = index;
+            }
+            _ => panic!("Taken in per-module free list"),
+        }
+    }
+}
+
+impl PoolingAllocationState {
+    /// Create the default state for this strategy.
+    pub(crate) fn new(strategy: PoolingAllocationStrategy, max_instances: usize) -> Self {
+        let ids = (0..max_instances).collect::<Vec<_>>();
+        match strategy {
+            PoolingAllocationStrategy::NextAvailable => PoolingAllocationState::NextAvailable(ids),
+            PoolingAllocationStrategy::Random => PoolingAllocationState::Random(ids),
+            PoolingAllocationStrategy::ReuseAffinity => PoolingAllocationState::ReuseAffinity {
+                free_list: ids,
+                per_module: HashMap::new(),
+                slot_state: (0..max_instances)
+                    .map(|i| SlotState::Empty { free_list_index: i })
+                    .collect(),
+            },
+        }
+    }
+
+    /// Are any slots left, or is this allocator empty?
+    pub(crate) fn is_empty(&self) -> bool {
+        match self {
+            &PoolingAllocationState::NextAvailable(ref free_list)
+            | &PoolingAllocationState::Random(ref free_list) => free_list.is_empty(),
+            &PoolingAllocationState::ReuseAffinity { ref free_list, .. } => free_list.is_empty(),
+        }
+    }
+
+    /// Internal: remove a slot-index from the global free list.
+    fn remove_free_list_item(
+        slot_state: &mut Vec<SlotState>,
+        free_list: &mut Vec<usize>,
+        index: usize,
+    ) {
+        let free_list_index = slot_state[index].free_list_index();
+        assert_eq!(index, free_list.swap_remove(free_list_index));
+        if free_list_index < free_list.len() {
+            let replaced = free_list[free_list_index];
+            slot_state[replaced].update_free_list_index(free_list_index);
+        }
+    }
+
+    /// Internal: remove a slot-index from a per-module free list.
+    fn remove_module_free_list_item(
+        slot_state: &mut Vec<SlotState>,
+        per_module: &mut HashMap<CompiledModuleId, Vec<usize>>,
+        id: CompiledModuleId,
+        index: usize,
+    ) {
+        let per_module_list = per_module.get_mut(&id).unwrap();
+        let per_module_index = slot_state[index].per_module_index();
+        assert_eq!(index, per_module_list.swap_remove(per_module_index));
+        if per_module_index < per_module_list.len() {
+            let replaced = per_module_list[per_module_index];
+            slot_state[replaced].update_per_module_index(per_module_index);
+        }
+        if per_module_list.is_empty() {
+            per_module.remove(&id);
+        }
```

Comment:
Maybe

```rust
debug_assert!(per_module.contains(&id));
debug_assert!(!per_module.get(&id).is_empty());
```

at the start of this function?

Reply:
The first is covered by the `get_mut()` at the top; I added a `.expect()` rather than `.unwrap()` to make any failure clearer though. Added the second.

Reply:
> The first is covered by the `get_mut()` at the top

FWIW, I don't mind having "redundant" assertions because, in some sense, debug assertions are *always* redundant, since we are checking things that we believe should always be true.